### ISSUE TITLE
Client application startup failure using Greenwich.SR1 with SCS 2.x service instances

### ISSUE THREAD

**Issue by csterwa** [pivotal-cf/spring-cloud-services-connector#124]
Original issue found by @pacphi 

Using the following branch that was created to update client applications to using Spring Cloud Greenwich.SR1 and Spring Boot 2.1:

https://github.com/Pivotal-Field-Engineering/devops-workshop/tree/upgrade-to-spring-boot-2.1.4.release

The client applications ran into failures on startup:

```
   2019-04-22T18:51:21.28-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:21.289  WARN 14 --- [           main] o.s.boot.actuate.endpoint.EndpointId     : Endpoint ID 'service-registry' contains invalid characters, please migrate to a valid format.
   2019-04-22T18:51:21.88-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:21.882  INFO 14 --- [           main] o.s.cloud.context.scope.GenericScope     : BeanFactory id=59fdf168-4b96-3fc7-810f-89ac7db619a9
   2019-04-22T18:51:21.89-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:21.893  INFO 14 --- [           main] eReconfigurationBeanFactoryPostProcessor : Skipping reconfiguration because cloud services already configured
   2019-04-22T18:51:22.36-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:22.364  INFO 14 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration' of type [org.springframework.transaction.annotation.ProxyTransactionManagementConfiguration$$EnhancerBySpringCGLIB$$3f825871] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
   2019-04-22T18:51:22.41-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:22.417  INFO 14 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'org.springframework.hateoas.config.HateoasConfiguration' of type [org.springframework.hateoas.config.HateoasConfiguration$$EnhancerBySpringCGLIB$$bf02a5a3] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
   2019-04-22T18:51:22.43-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:22.431  INFO 14 --- [           main] trationDelegate$BeanPostProcessorChecker : Bean 'org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration' of type [org.springframework.cloud.autoconfigure.ConfigurationPropertiesRebinderAutoConfiguration$$EnhancerBySpringCGLIB$$5b9c5b6e] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
   2019-04-22T18:51:23.29-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:23.298  INFO 14 --- [           main] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
   2019-04-22T18:51:23.35-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:23.353  INFO 14 --- [           main] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
   2019-04-22T18:51:23.35-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:23.353  INFO 14 --- [           main] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.17]
...
   2019-04-22T18:51:29.43-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:29.433  INFO 14 --- [           main] j.LocalContainerEntityManagerFactoryBean : Initialized JPA EntityManagerFactory for persistence unit 'default'
   2019-04-22T18:51:30.59-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:30.597  INFO 14 --- [           main] o.h.h.i.QueryTranslatorFactoryInitiator  : HHH000397: Using ASTQueryTranslatorFactory
   2019-04-22T18:51:31.25-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.250  WARN 14 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration$ConfigClientOAuth2Configurer': Unsatisfied dependency expressed through constructor parameter 0; nested exception is org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration' available: expected single matching bean but found 2: configClientOAuth2BootstrapConfiguration,io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration
   2019-04-22T18:51:31.25-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.253  INFO 14 --- [           main] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
   2019-04-22T18:51:31.26-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.262  INFO 14 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
   2019-04-22T18:51:31.27-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.272  INFO 14 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.
   2019-04-22T18:51:31.27-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.274  INFO 14 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.293  WARN 14 --- [           main] o.a.c.loader.WebappClassLoaderBase       : The web application [ROOT] appears to have started a thread named [spring.cloud.inetutils] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/jdk.internal.misc.Unsafe.park(Native Method)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.locks.LockSupport.park(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.LinkedBlockingQueue.take(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.ThreadPoolExecutor.getTask(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
   2019-04-22T18:51:31.29-0700 [APP/PROC/WEB/0] OUT  java.base@11.0.2/java.lang.Thread.run(Unknown Source)
   2019-04-22T18:51:31.32-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.326  INFO 14 --- [           main] ConditionEvaluationReportLoggingListener : 
   2019-04-22T18:51:31.32-0700 [APP/PROC/WEB/0] OUT Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT 2019-04-23 01:51:31.329 ERROR 14 --- [           main] o.s.b.d.LoggingFailureAnalysisReporter   : 
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT ***************************
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT APPLICATION FAILED TO START
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT ***************************
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT Description:
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT Parameter 0 of constructor in io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration$ConfigClientOAuth2Configurer required a single bean, but 2 were found:
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT     - configClientOAuth2BootstrapConfiguration: defined in URL [jar:file:/home/vcap/app/BOOT-INF/lib/spring-cloud-services-spring-connector-2.1.1.RELEASE.jar!/io/pivotal/spring/cloud/service/config/ConfigClientOAuth2BootstrapConfiguration.class]
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT     - io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration: defined in null
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT Action:
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT Consider marking one of the beans as @Primary, updating the consumer to accept multiple beans, or using @Qualifier to identify the bean that should be consumed
   2019-04-22T18:51:31.72-0700 [APP/PROC/WEB/0] OUT Exit status 1
   2019-04-22T18:51:31.72-0700 [CELL/SSHD/0] OUT Exit status 0
```

The same code currently works with Spring Cloud Finchley on `master` branch.

**Comment by habuma** [comment]
A few things I've learned thus far...

- If I put `@Primary` on the `ConfigClientOAuth2BootstrapConfiguration` class, it makes things work, although I think that's more of a workaround than a fix.
- I put some logging in place to see when `ConfigClientOAuth2BootstrapConfiguration` is created. In the `solutions/05/cloud-native-spring` sample, it gets created *TWICE*...once in the bootstrap context (which is correct) and once in the application's context (which shouldn't be).
- I tried the "Cook" sample app with the *exact* same dependencies (Boot 2.1.4, SC Greenwich.SR1, and my local build of connectors with the logging) and that class only gets created once, in the bootstrap context...which is correct. 
- Before putting my local connectors build in, I tried the "Cook" sample with Boot 2.1.4, SC Greenwich.SR1, and SC Connectors 2.1.1...and it worked fine.

This tells me that either there's something special about the `solutions/05/cloud-native-spring` project that causes that config to be loaded twice or that there's something amiss in some of the other SCS dependency autoconfig that's tricking it into doing that. Not sure yet...still digging. Just reporting what I've seen thus far.

**Comment by pacphi** [comment]
@habuma Mind posting your workaround snippet here? 

I'm thinking this has something to do with auto-config.  The bit below seems to indicate multiple `ConfigServicePropertySourceLocator` candidates are identified on class path. That is the first argument to the constructor for `ConfigClientOAuth2Configurer`. 

```
Parameter 0 of constructor in io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration$ConfigClientOAuth2Configurer required a single bean, but 2 were found:
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT     - configClientOAuth2BootstrapConfiguration: defined in URL [jar:file:/home/vcap/app/BOOT-INF/lib/spring-cloud-services-spring-connector-2.1.1.RELEASE.jar!/io/pivotal/spring/cloud/service/config/ConfigClientOAuth2BootstrapConfiguration.class]
   2019-04-22T18:51:31.33-0700 [APP/PROC/WEB/0] OUT     - io.pivotal.spring.cloud.service.config.ConfigClientOAuth2BootstrapConfiguration: defined in null
```

**Comment by habuma** [comment]
@pacphi It isn't even really a workaround as much as it is a hack. And it is in the connector code, not in the sample code. It involves putting `@Primary` on the `ConfigClientOAuth2BootstrapConfiguration` class. 

It definitely has something to do with auto-config, but context refresh has some part to play in it, too. And yes, I've noticed those log entries, but they're not saying what you say that they're saying. They're actually kinda misleading...it's not complaining that there are two `ConfigClientOAuth2Configurer`, but that there are two `ConfigClientOAuth2BootstrapConfiguration` beans...which, of course, aren't the parameters to that constructor.

A bit more backstory: I've hacked at this by completely REMOVING that constructor altogether. I didn't expect it would work correctly, but I expected to see a different error. But instead, I kept getting the same error even with no constructor in that class at all.

Moreover, based on logging I put into the connector code, I can see that (in the case of the workshop sample apps), `ConfigClientOAuth2BootstrapConfiguration` is created twice (which matches what the error says). Meanwhile when I use the latest Boot, Greenwich.SR1, and SC connectors with our "cook" example (https://github.com/spring-cloud-services-samples/cook) it works fine...and `ConfigClientOAuth2BootstrapConfiguration` is only created once. 

One more thing: I've started stripping down the workshop sample app to barebones, removing anything that might be a culprit and making it resemble the "cook" app. My hope is that at some point I'll remove or change something that makes it work and then I'll know what's at fault. But so far, it still fails...even though I've nearly removed everything and it's now a near-clone of "cook". But I'm still looking...bound to be some difference.

I'm growing less convinced that this is a connector bug and more convinced that it's something unique to this sample, either in the sample itself or something about it that uncovers a bug in Boot or Spring Cloud that the "cook" sample doesn't uncover.

**Comment by habuma** [comment]
I still don't know *WHY* this is happening, but... (and yes, this is weird)...

I can make it work by changing the base package name of the sample from `io.pivotal` to (I guess) anything else. I changed it to `io.pivotalx` and it works fine.

**Comment by habuma** [comment]
Okay...figured it out.

The sample project has a base package of `io.pivotal`, which is also the package that its main bootstrap class resides. The main class is annotated with `@SpringBootApplication`, which has an implicit `@ComponentScan` which scans that package (`io.pivotal.**`) and all of its subpackages. 

Meanwhile, the SCS connectors auto-config class resides in `io.pivotal.spring.cloud.service.config`, which happens to be a subpackage of `io.pivotal`. (See where this is going???)

When the app starts up, auto-configuration takes place and our auto-configuration class is loaded and evaluated in the bootstrap application context.

Then…when the application itself starts up, component-scanning starts scanning the `io.pivotal` package hierarchy and finds our autoconfig in `io.pivotal.spring.cloud.service.config` and creates and evaluates it a second time in the application context. This causes the application to fail at startup.

Simply repackaging the sample app so that it’s in a different hierarchy than `io.pivotal` keeps our auto-configuration from being evaluated twice and allows the app to start with no issues.

The remaining mystery: Supposedly this worked fine up until it was upgraded to Boot 2.1.4. But it should’ve broken in the same way on any version of Boot, I’d think. Unless the change in how bean overriding has something to do with it (likely).

In short, this is not a bug in SCS Connectors, but rather a conflict in the application's own component-scanning along with our choice of package names. And, possibly exposed by changes in Spring Boot to how bean override works.

Closing with the suggestion of either renaming the sample projects' base package or alternatively excluding `io.pivotal.spring.cloud.service.config` from component-scanning. (Although, the former is the preferred solution.)

**Comment by pacphi** [comment]
Thanks @habuma for the thorough research! Repackaging the workshop to have a base package of `io.pivotal.cns`.
